lib/datasource.js

```diff
@@ -365,36 +365,60 @@ DataSource.prototype.setup = function(name, settings) {
     }
   }
 
-  dataSource.connect = function(cb) {
-    var dataSource = this;
-    if (dataSource.connected || dataSource.connecting) {
-      process.nextTick(function() {
-        cb && cb();
-      });
+  dataSource.connect = function(callback) {
+    var self = this;
+    if (this.connected) {
+      // It's already connected or being connected
+      // Do NOT run loadFromFile again as it will override the cache
+      if (callback) process.nextTick(callback);
       return;
     }
-    dataSource.connecting = true;
-    if (dataSource.connector.connect) {
-      dataSource.connector.connect(function(err, result) {
-        if (!err) {
-          dataSource.connected = true;
-          dataSource.connecting = false;
-          dataSource.emit('connected');
-        } else {
-          dataSource.connected = false;
-          dataSource.connecting = false;
-          dataSource.emit('error', err);
-        }
-        cb && cb(err, result);
-      });
-    } else {
+    if (typeof dataSource.connector.connect !== 'function') {
+      // Connector doesn't have the connect function
+      // Assume no connect is needed
       process.nextTick(function() {
-        dataSource.connected = true;
-        dataSource.connecting = false;
-        dataSource.emit('connected');
-        cb && cb();
+        self.connected = true;
+        self.connecting = false;
+        self.emit('connected');
+        callback && callback();
```

Comment:
Instead of checking whether `callback` is all over the place, it's better to move this check to the top of the method.


```js
callback = callback || function() {};
```

We should consider returning a Promise when no callback was specified, but that's probably out of scope of this patch.